packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,564 @@
+   */
+  async #handleAccountAssetListUpdated(
+    event: AccountAssetListUpdatedEventPayload,
+  ) {
+    this.#assertControllerMutexIsLocked();
+
+    const assetsToUpdate = event.assets;
+    let assetsForMetadataRefresh = new Set<CaipAssetType>([]);
+    for (const accountId in assetsToUpdate) {
+      if (hasProperty(assetsToUpdate, accountId)) {
+        const newAccountAssets = assetsToUpdate[accountId];
+        if (
+          newAccountAssets.added.length !== 0 ||
+          newAccountAssets.removed.length !== 0
+        ) {
+          const { added, removed } = newAccountAssets;
+          const existing = this.state.accountsAssets[accountId] || [];
+          const assets = new Set<CaipAssetType>([
+            ...existing,
+            ...added.filter((asset) => isCaipAssetType(asset)),
+          ]);
+          for (const removedAsset of removed) {
+            assets.delete(removedAsset);
+          }
+          assetsForMetadataRefresh = new Set([
+            ...assetsForMetadataRefresh,
+            ...assets,
+          ]);
+          this.update((state) => {
+            state.accountsAssets[accountId] = Array.from(assets);
+          });
+        }
+      }
+    }
+    // Trigger fetching metadata for new assets
+    await this.#refreshAssetsMetadata(Array.from(assetsForMetadataRefresh));
+  }
+
+  /**
+   * Checks for non-EVM accounts.
+   *
+   * @param account - The new account to be checked.
+   * @returns True if the account is a non-EVM account, false otherwise.
+   */
+  #isNonEvmAccount(account: InternalAccount): boolean {
+    return (
+      !isEvmAccountType(account.type) &&
+      // Non-EVM accounts are backed by a Snap for now
+      account.metadata.snap !== undefined
+    );
+  }
+
+  /**
+   * Handles changes when a new account has been added.
+   *
+   * @param account - The new account being added.
+   */
+  async #handleOnAccountAdded(account: InternalAccount): Promise<void> {
+    if (!this.#isNonEvmAccount(account)) {
+      // Nothing to do here for EVM accounts
+      return;
+    }
+    this.#assertControllerMutexIsLocked();
+
+    // Get assets list
+    if (account.metadata.snap) {
+      const assets = await this.#getAssetsList(
+        account.id,
+        account.metadata.snap.id,
+      );
+      await this.#refreshAssetsMetadata(assets);
+      this.update((state) => {
+        state.accountsAssets[account.id] = assets;
+      });
+    }
+  }
+
+  /**
+   * Handles changes when a new account has been removed.
+   *
+   * @param accountId - The new account id being removed.
+   */
+  async #handleOnAccountRemovedEvent(accountId: string): Promise<void> {
+    // Check if accountId is in accountsAssets and if it is, remove it
+    if (this.state.accountsAssets[accountId]) {
+      this.update((state) => {
+        // We are not deleting the assetsMetadata because we will soon make this controller extends StaticIntervalPollingController and update all assetsMetadata once a day.
```

Comment:
```suggestion
        // TODO: We are not deleting the assetsMetadata because we will soon make this controller extends StaticIntervalPollingController
        // and update all assetsMetadata once a day.
```